jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -266,33 +270,76 @@ public String toString()
         private final SynchronousQueue<Runnable> _task = new SynchronousQueue<>();
-        private boolean _starting = true;
+
+        // Reserved Thread state is an integer that is either one of the special values
+        // below; 0 for running normally or a small positive value that counts the number
+        // of consecutive offer misses.
+        private static final int PENDING = Integer.MAX_VALUE;
+        private static final int RUNNING = 0;
+        private static final int STOPPED = -1;
+        private static final int MISSED_OUT = -2;
+        private static final int IDLED_OUT = -3;
+        private final AtomicInteger _state = new AtomicInteger(PENDING);
+        private Thread _thread;
 
         public boolean offer(Runnable task)
         {
             if (LOG.isDebugEnabled())
-                LOG.debug("{} offer {}", this, task);
+                LOG.debug("{} offer task={} {}", this, task, ReservedThreadExecutor.this);
 
             try
             {
-                _task.put(task);
-                return true;
+                // Offer the task to the SynchronousQueue, but without blocking or even yielding.
+                // The intent of RTE is to support EPC strategy, thus if we block or even yield here then we give
+                // an opportunity for the caller to switch CPUs, defeating the intent of EPC.
+                if (_task.offer(task))
+                    return true;
+
+                // check state and potential reason for the miss
+                while (true)
+                {
+                    int state = _state.get();
+                    // Are we not in a state that should have been offered to?
+                    if (state < 0 || state == PENDING)
+                    {
+                        LOG.warn("ReservedThread.offered to: {}", this);
+                        return false;
+                    }
+
+                    // Have we missed too often?
+                    if (state >= MAX_CONSECUTIVE_MISSES)
+                    {
+                        if (_state.compareAndSet(state, MISSED_OUT))
+                            continue;
+                        LOG.warn("ReservedThread.offer max missed: {}", this);
+                        return false;
+                    }
+
+                    // increment the misses
+                    if (_state.compareAndSet(state, state + 1))
+                        break;
+                }
             }
             catch (Throwable e)
             {
-                LOG.ignore(e);
-                _size.getAndIncrement();
-                _stack.offerFirst(this);
-                return false;
+                LOG.warn(e);
             }
+
+            // We failed to offer the task, so put this thread back on the stack in last position to give it time to arrive.
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} offer missed {}", this, task, ReservedThreadExecutor.this);
+            _count.add(0, -1);
```

Comment:
This can also make `size` go below 0 and trigger the "stop executing anything on the `ReservedThreads`" bug.